Feature/task 2

PetrSvirak opened the pull request.

vaclavm committed:
Merge branch 'develop' of http://github.com/vaclavm/kentico-onboarding-cs into develop

vaclavm committed:
KA-274 Create ToDo controller with dummy items + test

Controller returns dummy ToDo objects. Tests are written for the responsive of the controller. Controller is with versoning
Removed MVC/Web application assymply from web.config.

vaclavm committed:
KA-274 Change test to Constraint model

vaclavm committed:
KA-275 Introduce dummy repository

Model moved from API to Contracts project. Moved repository to Repository project. Update tests.

vaclavm committed:
KA-275 Introduce Unity DI and NSubstitute to test repository pattern

Repository is still public. Testing only API.

vaclavm committed:
KA-276 Update constraint resolve to work with versioning

vaclavm committed:
KA-275 Change dependency injection to proper usage

Using DI with interfaces and internal services. Created URL helper to get url after post.
Not working right now due to problems with httpmessagerequest passing to service.

vaclavm committed:
KA-275 Register HttpRequestMessage correctly

vaclavm committed:
KA-274 Simplify controller and update tests

Change to lambda notation. Update route in post.

vaclavm committed:
KA-274 Correct tests

vaclavm committed:
KA-274 Make ToDo comparer lazy

vaclavm committed:
KA-274 Change to generic controller test

Add extension for ApiController class to execute actions with none cancellation token. Simplify the tests to use it.

vaclavm committed:
KA-275 Merge changes from feature/task-1

Updated tests with extensions to equalTo and controller actions.

vaclavm committed:
KA-275 Update files after merge

Update ToDoModel to ToDo as it was changed. Change tests to work after merge. Rebuild whole solution.

vaclavm committed:
KA-275 Change url location helper to use url helper

Using url helper to get proper route from WebApi. Substitute the helper in test.

vaclavm committed:
KA-275 Change DI to register HttpRequestMessage

Helper is now initialized with UrlHelper as HttpRequest message is registred

vaclavm committed:
KA-275 Simplify files

vaclavm committed:
KA-275 Move DI setup from WebApiConfig to a new file

vaclavm committed:
KA-275 Update folder structure

Create src and test folders and moved project inside the folders.

vaclavm committed:
KA-275 Update config files

Add new config for JSON camelCase formatter. Rename other configurations.

vaclavm committed:
KA-275 Add private constructor in comparer

vaclavm committed:
KA-275 Refactor controller tests to use statis items field

vaclavm committed:
KA-275 Change repository to return IEnumerable instead of List

vaclavm committed:
KA-275 Move dependency injection to separate project

Create wrapper for Unity DI to separate other projects from Unity library.
Moved helpers from WebApi to separate project and change them to services.
As there is a wrapper now, we can test it and if anything happens with Unity, recreate it with different functionality.

vaclavm committed:
KA-275 Add get route path to location service

Location service now has route path for get. It can properly return new resource location.

vaclavm committed:
KA-275 Create RouteService to maintain routes for API

The service maintain routes for the API and can be mocked in tests. As location service is outside API, the route service helps it to get correct route name when creating new location.

vaclavm committed:
KA-275 Play with dependency injection

Separating DI from HTTP to have clean container.

vaclavm committed:
KA-275 Move DI for API to separate project

DI for API moved so the API doesn't know about implementation (UNITY). All other projects know about DI and needs it to work correctly.
Renamed Wrapper to Container to better describe the funcionality. Also create bootstraper for dependency resolver, calling it from API.

vaclavm committed:
KA-275 Make correct disposable method in DI resolver

https://kentico.atlassian.net/wiki/spaces/KID/pages/86769986/IDisposable

vaclavm committed:
Merge branch 'develop' of http://github.com/vaclavm/kentico-onboarding-cs into develop

vaclavm committed:
KA-275 Merge from develop

vaclavm committed:
KA-275 Delete previous files after merging develop

vaclavm committed:
KA-275 Create test for Api.Services

vaclavm committed:
KA-275 Restore NuGet packages

Solve problem with NuGet packages after dividing the application to src and test folders

vaclavm committed:
KA-275 Add APIVersioning reference

PetrSvirak commented on ToDoList/src/ToDoList.API/App_Start/DependencyInjectionConfig.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        public static void Register(HttpConfiguration config)
+        {
+            var diBootstrap = new DependencyBootstraper();
```
Omit [abbreviations](https://www.urbandictionary.com/define.php?term=D.I.#272058), please - `bootstrapper` is ok as would `dependencyInjectionBootrapper` (since the class is called "…Bootstrapper" I would not go with "bootstrap")

PetrSvirak commented on ToDoList/src/ToDoList.API.DependencyInjection/DependencyBootstraper.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace ToDoList.API.DependencyInjection
+{
+    public class DependencyBootstraper
```
There should be [double `p`](https://study.com/academy/lesson/english-spelling-rules-for-doubling-dropping-letters.html) in `bootstrapper`

PetrSvirak commented on ToDoList/src/ToDoList.API/App_Start/DependencyInjectionConfig.cs:
```diff
@@ -0,0 +1,20 @@
+        {
+            var diBootstrap = new DependencyBootstraper();
+            diBootstrap.RegisterDependencies();
+            diBootstrap.RegisterApiSingleton<IRoutesService, RoutesHelper>();
```
I'm just wondering here whether it is not an implementation detail of the `bootstrapper` that it calls `RegisterDependencies` and `RegisterSingleton` - I mean, wouldn't code like this be a bit more readable:
````
var resolver = DependencyBootstrapper.CreateWebApiResolver(RoutesHelper)
config.DependencyResolver = resolver;
````
````
public static IDependencyResolver CreateWebApiResolver(IRoutesService routeService)
  => new DependencyBootstraper(new Container())
        .Register<Repository.DependencyRegister>(_container)
        .Register<Services.DependencyRegister>(_container)
        .Register<IRoutesService>(routeService);
````
Note that:
* I passed an instance of a class that implements `IRoutesService` since `singleton` is typically instantiable anytime during application's lifetime. It seems a bit binding to expect all (API) singletons to have parameterless constructors; on the other hand, being instantiable at the application start might prove as binding if the instance is costly and required later in the lifetime or conditionally - both problems have their respective solutions, of course. Have a look at [this comment](https://github.com/vaclavm/kentico-onboarding-cs/pull/6/files#diff-5c05d1792f39fb09919304131f948f0cR5) and the way MS sorts something similar out in .NET Core.
* I passed `Container` instance into the `DependencyBootstraper` constructor, so it can be tested (see below).

PetrSvirak replied:
There is a test that might prove very useful, especially if collaborating with other developers that do not know the project by heart. Since you have contracts, you can easily reflect a set of interfaces that need to be implemented in your (Web API) project. Basically, they are all interfaces from `.Contracts` assembly except `IDependencyRegister`. A test that calls `DependencyBootstrapper` and examines whether it is possible to obtain instances of all the contracts is what I miss here.

vaclavm replied:
I have refactored it and add the test as you proposed :).

PetrSvirak commented on ToDoList/src/ToDoList.Contracts/IDependencyRegister.cs:
```diff
@@ -0,0 +1,9 @@
+﻿using ToDoList.DependencyInjection;
```
What a mess 😛 `Contracts` should not require any other assembly and you do not use your `IDependencyConstruct` at all 😉

PetrSvirak commented on ToDoList/src/ToDoList.Contracts/IDependencyRegister.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public interface IDependencyRegister
+    {
+        void Register(Container container);
```
Please, make the code fluent where possible - it makes code shorter and less blocky.

PetrSvirak commented on ToDoList/src/ToDoList.Contracts/IDependencyConstruct.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace ToDoList.Contracts
+{
+    public interface IDependencyConstruct
```
It is more of an `IContainer`, isn't it?

PetrSvirak replied:
It is also outdated (when compared to the `Container` class).

PetrSvirak replied:
And there are some unused usings up there.

vaclavm replied:
Deleted, it was a relic

PetrSvirak commented on ToDoList/src/ToDoList.Contracts/IDependencyRegister.cs:
```diff
@@ -0,0 +1,9 @@
+﻿using ToDoList.DependencyInjection;
+
+namespace ToDoList.Contracts
```
Consider moving `DI` and `bootstrap` related contracts to folders. Something like
* `Bootstrap`
  * `IDependencyRegister` (or `IBootstrap`)
  * `IBootstrapper`
* `DependencyInjection`
  * `IContainer`
  * `LifetimeManager` (consider plural here)

-- or --

* `DependencyInjection` (or `Bootstrap`)
  * `IDependencyRegister` (or `IBootstrap`)
  * `IBootstrapper`
  * `IContainer`
  * `LifetimeManager` (consider plural here)

vaclavm replied:
In the end, this is not needed. The DI was completely moved from ToDoList.Contracts to ToDoList.DependencyInjection

PetrSvirak commented on ToDoList/src/ToDoList.API/App_Start/JsonConfig.cs:
```diff
@@ -0,0 +1,14 @@
+        public static void Register(HttpConfiguration config)
+        {
+            config.Formatters.JsonFormatter.SerializerSettings.ContractResolver = new CamelCasePropertyNamesContractResolver();
+            config.Formatters.JsonFormatter.UseDataContractJsonSerializer = false;
```
I'm not a friend of local variables for properties, however, this really is long and has half of the path common, so please consider something like `var formatter = config.Formatters.JsonFormatter`…

PetrSvirak commented on ToDoList/src/ToDoList.API/App_Start/RouteConfig.cs:
```diff
@@ -1,16 +1,21 @@
+        }
 
-            // Web API constrains
+        private static DefaultInlineConstraintResolver CreateConstrainResolver()
```
Really nice :)

PetrSvirak replied:
Method should return `IInlineConstraintResolver` - but that is just a nitpicking detail 👼

PetrSvirak commented on ToDoList/src/ToDoList.API/Controllers/ToDosController.cs:
```diff
@@ -0,0 +1,56 @@
+    [ApiVersion("1.0")]
+    [RoutePrefix("api/v{version:apiVersion}/todos")]
+    [Route("")]
+    public class ToDosController : ApiController
```
Your controller looks pretty straight-forward and single-minded - perfect :)

PetrSvirak commented on ToDoList/src/ToDoList.API/Helpers/RoutesHelper.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace ToDoList.API.Helpers
+{
+    internal class RoutesHelper : IRoutesService
```
`EMS` habits seem to manifest here a bit 😆

Have a look at `Options` (or `Config(uration)`) pattern as [Microsft presents it for .NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options) - it is closely related to both usage I propose (regards container method) and name - this here really is just an object caring data around. It is not exactly `model` (DTO) though, since it needs to be instantiated in a special way (e.g. it reads its values from internal constants it incorporates rather than having them set by an application logic as would be expected from a true [POCO](https://en.wikipedia.org/wiki/POCO)). 

However, in this comment here, I'm on the names - neither `Helper` nor `Service` suits in my eyes. There is definitely more, however, I see three options here: `I((Web|Rest)Api)Routes` or `IRouteConfig(uration)` or `I((Web|Rest)Api)Options` (this one has broader responsibility)

vaclavm replied:
IWebApiRoutes suits me :)

PetrSvirak commented on ToDoList/src/ToDoList.API/Global.asax.cs:
```diff
@@ -0,0 +1,14 @@
+            GlobalConfiguration.Configure(RouteConfig.Register);
+            GlobalConfiguration.Configure(DependencyInjectionConfig.Register);
+            GlobalConfiguration.Configure(JsonConfig.Register);
+        }
```
Looks really nice now, thanks.

PetrSvirak commented on ToDoList/src/ToDoList.API.DependencyInjection/DependencyResolver.cs:
```diff
@@ -0,0 +1,65 @@
+                throw new ArgumentNullException(nameof(container));
+            }
+            _unityContainer = container;
+        }
```
R# actually offers pretty nice simplification with C# 7: `_unityContainer = container ?? throw new ArgumentNullException(nameof(container));` – it based on fact that plenty language were made into expression lately, so was [`throw`](http://structuredsight.com/2016/09/01/c-7-additions-throw-expressions/)…

PetrSvirak commented on ToDoList/src/ToDoList.API.DependencyInjection/DependencyResolver.cs:
```diff
@@ -0,0 +1,65 @@
+using Unity;
+using Unity.Exceptions;
+
+namespace ToDoList.API.DependencyInjection
```
It is preferable to leave only "entry-point" classes (if any) in the root of an assembly. While it is implementation of the `IDependencyRegister` in most other assemblies, here it should be a `DependencyBootstraper` here (or if it may be "by the book", there could be a static `Bootstraper` class with single method creating an instance of `DependencyBootstraper` and everything)

vaclavm replied:
I am not sure what do you mean, I will ask in person.

PetrSvirak commented on ToDoList/ToDoList.sln:
```diff
@@ -3,9 +3,21 @@ Microsoft Visual Studio Solution File, Format Version 12.00
 VisualStudioVersion = 15.0.26228.9
 MinimumVisualStudioVersion = 10.0.40219.1
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ToDoList.API", "ToDoList.API\ToDoList.API.csproj", "{CFFC7437-BF3E-4BE7-A5CB-00EC5C111073}"
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ToDoList.API", "src\ToDoList.API\ToDoList.API.csproj", "{CFFC7437-BF3E-4BE7-A5CB-00EC5C111073}"
```
Could you please move projects to solution folders mimicking the "real" folder structure?

vaclavm replied:
I am not sure what do you mean, I will ask in person.

PetrSvirak commented on ToDoList/src/ToDoList.API.DependencyInjection/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,36 @@
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("ToDoList.API.DependencyInjection")]
```
This was first assembly info with `API` in title, so the comment goes here 👼 
Can you please rename all `API` projects to `Api` – it is essential in (default) namespaces, but changing it everywhere does not seem to be that extra work and looks more consistent…

PetrSvirak commented on ToDoList/src/ToDoList.API.DependencyInjection/DependencyBootstraper.cs:
```diff
@@ -0,0 +1,35 @@
+            Register<Services.DependencyRegister>(_container);
+        }
+
+        public void RegisterApiSingleton<TFrom, TTo>()
```
As mentioned at [this comment](https://github.com/vaclavm/kentico-onboarding-cs/pull/6/files#diff-edee51a4cbe7f75e88f4ebb21a636091R15), there might be just `TFrom` generic and a `TFrom` or `Func<TFrom>` as a parameter…

vaclavm replied:
Solve by registering instance instead of singleton

PetrSvirak commented on ToDoList/src/ToDoList.API.Services/DependencyRegister.cs:
```diff
@@ -0,0 +1,23 @@
+using ToDoList.Contracts.Services;
+using ToDoList.DependencyInjection;
+
+namespace ToDoList.API.Services
```
It is kinda of a custom to create a folder based on `.Contracts`' folder system in Cloud - there will be `Services` folder in the `Services` assembly and namespace goes to ToDoList.API.Services.Services, true, however, it is only visible in the assembly itself (since the implementation is `internal`) and flat structure is not maintainable in large scale. It also encourages most developers to introduce multiple folders with less generic names in the contract (and therefore implementing assemblies) and that makes the code (structure) more navigable…

The root of the assembly is then reserved for the "entry" point of the assembly - implementation of the `IDependencyRegister` in most of the cases. You can see the similar pattern in the WebApi itself - there is `web.config` and `package.config` - both non-code configuration files for respective services the project is using (or is used by) and then there is the `global.asax(.cs)` that server as the entry point for IIS. That's it, everything else is organized in folders…

PetrSvirak commented on ToDoList/src/ToDoList.DependencyInjection/Container.cs:
```diff
@@ -0,0 +1,39 @@
+                case LifetimeManager.Hierarchical:
+                    _container.RegisterType<TFrom, TTo>(new HierarchicalLifetimeManager());
+                    break;
+                default:
```
Consider stating the `LifetimeManager.Singleton` explicitly and throwing an exception on `default` (this will occur when a new value gets added to the enum yet this method is not updated) – that is considered a best practice (in C# and in JS as well)…

PetrSvirak replied:
It would also be nice to have tests for the switch, but that is kind of nitpicking of me (yet if you used TDD, you would end up with some, so I at least felt the urge to mention them).

PetrSvirak commented on ToDoList/test/ToDoList.API.Tests/Controllers/ToDosControllerTests.cs:
```diff
@@ -0,0 +1,126 @@
+            var expectedToDo = _toDoList.ElementAt(itemIndex);
+
+            _toDoRepositorySubstitute.AddToDoAsync(expectedToDo).Returns(expectedToDo);
+            _urlLocationServiceSubstitute.GetNewResourceLocation(expectedToDo.Id).Returns($"todos/{expectedToDo.Id}");
```
Consider moving the `string` into the arrange part - it's 50:50 since the string repeats twice, but constant mustn't add to readability…

PetrSvirak commented on ToDoList/test/ToDoList.API.Tests/Controllers/ToDosControllerTests.cs:
```diff
@@ -0,0 +1,126 @@
+            var response = await _controller.ExecuteAction(controller => controller.PutToDoAsync(expectedToDo.Id, expectedToDo));
+
+            // Assert
+            Assert.That(response.StatusCode, Is.EqualTo(HttpStatusCode.NoContent), $"Expecting status code NoContent, but is {response.StatusCode}");
```
Hm, thinking about it, since you return `NoContent`, you should at least check that the `repository` method was called with correct parameters… (same applies to the `Delete` method)

PetrSvirak commented on ToDoList/test/ToDoList.API.Services.Tests/UrlLocationServiceTests.cs:
```diff
@@ -0,0 +1,48 @@
+        private IRoutesService _routesService;
+        private UrlLocationService _locationService;
+
+        private const string WhatsUp = "whatsUp";
```
It is a `RouteName`, right? 😛

PetrSvirak commented on ToDoList/test/ToDoList.API.Services.Tests/UrlLocationServiceTests.cs:
```diff
@@ -0,0 +1,48 @@
+        public void SetUp()
+        {
+            var configuration = new HttpConfiguration();
+            configuration.Routes.MapHttpRoute("WhatsUp", $"{WhatsUp}/{{id}}", new { id = RouteParameter.Optional });
```
Please, add something to follow the `id` to prevent code that only suffixes what-ever path with `id`…

PetrSvirak commented on ToDoList/test/ToDoList.API.Services.Tests/UrlLocationServiceTests.cs:
```diff
@@ -0,0 +1,48 @@
+        public void GetNewResourceLocation_ReturnsCorrectLocation()
+        {
+            // Arrange
+            string location = "WhatsUp";
```
I wouldn't hesitate to use the constant here, but I'm open to discussion.

PetrSvirak commented on ToDoList/test/ToDoList.API.Services.Tests/UrlLocationServiceTests.cs:
```diff
@@ -0,0 +1,48 @@
+            var result = _locationService.GetNewResourceLocation(guid);
+
+            // Assert
+            Assert.That(result, Is.EqualTo($"/{WhatsUp}/{guid}"), $"The result location {result}, should be {WhatsUp}{guid}");
```
You miss slashes in the message; but this information can be read from `AssertException` itself - maybe change to message to state something like "`guid` is expected to be present in the middle of the URL" - there is actually xUnit used in Cloud where there is no message at all, but I like NUnit's ability to add more context for the developer reading the test and trying to figure out what it means that it failed (EMS stigma I guess 😆)…

PetrSvirak reviewed (commented):
On my, since I created the PR, I cannot "request changes" - my bad… Anyway, please change the following things. You've done quite some job on isolating `Unity` and while I agree it is kind of overkill, everything starts to fit in nicely. Hopefully, you enjoyed coding at least as much as I enjoyed the review 😃

vaclavm committed:
KA-275 Refactor DependencyBootstrapper to testable version

DependencyBootstrapper changed to return IDependencyResolver in public static method. Container is now part on constructor and the initialization of instances can be tested.

vaclavm committed:
KA-275 Clean Contracts projects

Move IDependencyRegister to DI project, delete IDependencyConstruct

vaclavm committed:
KA-275 Clean up Configs

vaclavm committed:
KA-275 Rename IRoutesService to IWebApiRoutes

vaclavm committed:
KA-275 Make code simplier by expression

vaclavm committed:
KA-275 Rename ToDoList.API to ToDoList.Api

vaclavm committed:
KA-275 Rename ToDoList.API.DependencyInjection to ToDoList.Api.DependencyInjection

vaclavm committed:
KA-275 Rename ToDoList.API.Services to ToDoList.Api.Services

vaclavm committed:
KA-275 Rename ToDoList.API.Tests to ToDoList.Api.Tests

vaclavm committed:
KA-275 Merging from origin

vaclavm committed:
KA-275 Rename ToDoList.API.Services.Tests to ToDoList.Api.Services.Tests

vaclavm committed:
KA-275 Change folders name from ...API... to ...Api...

vaclavm committed:
KA-275 Update travis.yml with new paths

vaclavm committed:
KA-275 Update packages

vaclavm committed:
KA-275 Rename project with API to Api

Using git mv due to Windows :-/

vaclavm committed:
KA-275 Update project with correct path to packages

vaclavm committed:
KA-275 Update travis.yml to correctly build

https://stackoverflow.com/questions/47964233/asp-net-mvc-project-building-starts-failing-on-travis-ci

vaclavm committed:
KA-275 Update dependencies

vaclavm committed:
KA-275 Change in tests

Testing method calling test, delete dummy test, cosmetic changes.

vaclavm committed:
KA-275 Create DependencyInjection folders in projects

vaclavm committed:
KA-275 Add folder to DependencyInjection

vaclavm committed:
KA-275 Add virtual folders - test, src - to solution

vaclavm commented:
I have decided to leave IDependencyRegister in ToDoList.DependencyInjection. Also, I have decided not to create IContainer for my Container class in DependencyInjection project.
There are multiple reasons for that, the main one is simplicity and clear definition of ToDoList.DependencyInjection project. Who wants to use it, needs to use IDependencyRegister to register its services so it does not make sense to move it into ToDoList.Contracts.
The container is also only one that registers services and needs to be public, so it can be used for application, therefore interface would only duplicate code.

vaclavm commented:
I have decided to leave IDependencyRegister in ToDoList.DependencyInjection. Also, I have decided not to create IContainer for my Container class in DependencyInjection project.
There are multiple reasons for that, the main one is simplicity and clear definition of ToDoList.DependencyInjection project. Who wants to use it, needs to use IDependencyRegister to register its services so it does not make sense to move it into ToDoList.Contracts.
The container is also only one that registers services and needs to be public, so it can be used for application, therefore interface would only duplicate code.

PetrSvirak commented:
I'm sorry, but I'm afraid you get it all wrong. There is absolutely no reason for any assembly to reference other projects than `Contracts`. The only exception is the assembly that actually registers other assemblies' contracts, of course. 

The `Container` does not need to be public at all since you only need its interface in all registration methods. And these methods are defined by the `IDependencyRegister` interface - if you put these into one assembly - `Contracts` - it will just click together. The API (if that is what you mean by "application") does not need the container at all as well - it expects the `IDependencyResolver` - it does not matter to it in what way you provide it, whether there is a `UnityContainer` or a `Dictionary` behind your implementation. 

The design is much clearer when you separate contracts and their implementation (I believe there was enough told about that). Stop looking at the DependencyInjection (assembly) like being extraordinary and different from other contracts you implement. There is somebody - `Contracts` - that promises that there will be an implementation of  `IContainer` with some methods for type registrations. In what way is it different to the `IRepostiory` that promises an implementation will be available that will allow (controller) to access a database? All other assemblies must implement the bootstrapping class and this class is defined by an interface with a method 

that promises the `IContainer` is provided and assemblies can register their contracts with it. In what way is it different to a service that is consumed by a controller yet requires other service or repository in its constructor?

One more thing, interfaces do not contain code so they cannot duplicate it 😛 While I understand where you are heading, please try to see the interfaces as contracts or promises and implementations as the fulfillment of such contracts or promises. Contracts allow you to easily change technology and isolate dependencies from their spread across the application. Writing functions twice (where IDE writes it once for me) is a prince I'm willing to pay.

![dependency separation](https://user-images.githubusercontent.com/15132110/37165587-4ddab50e-22fd-11e8-9e00-07c395fd98cd.png)

PetrSvirak commented on ToDoList/src/ToDoList.Api/ToDoList.Api.csproj:
```diff
@@ -163,15 +163,6 @@
-    </PropertyGroup>
-    <Error Condition="!Exists('..\..\packages\Microsoft.Net.Compilers.1.3.2\build\Microsoft.Net.Compilers.props')" Text="$([System.String]::Format('$(ErrorText)', '..\..\packages\Microsoft.Net.Compilers.1.3.2\build\Microsoft.Net.Compilers.props'))" />
-    <Error Condition="!Exists('..\..\packages\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.1.0.3\build\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.props')" Text="$([System.String]::Format('$(ErrorText)', '..\..\packages\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.1.0.3\build\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.props'))" />
-    <Error Condition="!Exists('..\..\packages\Microsoft.Net.Compilers.2.6.1\build\Microsoft.Net.Compilers.props')" Text="$([System.String]::Format('$(ErrorText)', '..\..\packages\Microsoft.Net.Compilers.2.6.1\build\Microsoft.Net.Compilers.props'))" />
```
I don't like the fact that you removed even conditions for the new packages. These should be preserved since results of build process are unpredictable if these packages are missing…

vaclavm replied:
I have added back the relevant errors (with proper version) :)

PetrSvirak commented on ToDoList/src/ToDoList.Api.DependencyInjection/DependencyResolver.cs:
```diff
@@ -1,9 +1,12 @@
 using Unity;
 using Unity.Exceptions;
 
+[assembly: InternalsVisibleTo("ToDoList.API.DependencyInjection.Tests")]
```
Since it is an assembly-wide attribute, please, place it into the `AssemblyInfo.cs`.

PetrSvirak reviewed (commented):
The comment above is my main concern of course, but there are still 2 other details. I admit I did not see your solution as whole just yet, but that is only because I did see it during the review before and I will see once you are finished with the DI.

vaclavm commented:
Ok, I get where you heading. It is a slight design shift from mine, as I tried to have DI as a separate project without reference to any other ToDoList projects.

vaclavm committed:
KA-275 Refactor DependencyInjection

I have created IContainer in Contracts project, it is used in DI project as Container is now internal. Also moved IDependencyRegister was moved to Contracts project, co DI project could be separated from other project.
Now it only communicates with API.DI, the structure is more clear.

vaclavm committed:
KA-275 Move InternalsVisibleTo to AssemblyInfo

vaclavm committed:
KA-275 Add Error Conditions to project file

vaclavm commented on ToDoList/src/ToDoList.Contracts/DependencyInjection/IContainer.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    public interface IContainer
+    {
+        object GetContainer();
```
Using object here was the only way I found how to separate implementation of DI (IUnityContainer) from other parts of the application (like ToDo.Contracts -> IDependencyRegister, where IContract is used).
I have tried to go with IContainer<T> ... or T GetContainer<T>, but every time I was forced to add IUnityContainer somewhere in the other project (like I mentioned - in ToDo.Contracts). Therefore using object seems like a right solution as I can determine in UnityDependencyResolver if the returned container has the correct type.

PetrSvirak replied:
That's because you are trying to wrap the `unity container` only partially - DI has two parts:
* _registration_ – that's what `IContainer` already does
* _resolution_ – that's what you need the `GetContainer`* method for since `IContainer` does not provide any way to resolve registered types

We later played around with the idea of separating these two parts into separate interfaces with Miro. However, in general, DI frameworks will usually provide you with the container exactly to allow you resolving on it (e.g. [`InjectionFactory` in `Unity`](https://msdn.microsoft.com/en-us/library/ff661868(v=pandp.51).aspx)). On the other hand, implementation of `IDependencyResolver` is the only place that needs the container for its resolving capabilities in your case and needs to resolve types in pretty loosely typed way (it expects an `object` to be returned while generic would ensure that the returned instance is an instance of the expected type or its successor as soon as during build). 

I don't believe you can go the very same way as Miro, since your solutions are approached from bit different angles. You might still create an interface allowing resolution (basically just 2 methods – `ResolveType` and `ResolveTypes`  – that would propagate towards wrapped `IUnityContainer`) and add another method to the `ContainerFactory`. You would have to make it create only one instance of the container at all times (there is the `Lazy` class to the rescue of course). This leaves two ways the container is used separate, however, it expects you to know that there is such factory that contains such method that returns such interface.

You could also introduce the interface described above and change the `GetContainer`* method's signature to return it (method would return `this`). This, however, allows anybody to obtain the resolution-enabled interface at any time (especially in one of `IDependencyRegister` implementations) – there is then no reason to separate the two interfaces. Mind you, we are still speaking about the very same instance of the container wrapper what so ever, so it would be better to simply extend `IContainer` with `ResolveType` and `ResolveTypes` methods (and "stay true and open" about the `IContainer`).

\* – the name is not clear now, you should've used something like `GetInnerContainer` or something that would explain it is not the same container (or its copy) that is being returned from the method.

vaclavm committed:
KA-275 Create resolution methods in Container

Separate Unity from API.DI only to DI. The UnityContainer is now fully wrapped by mine Container.
